server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java

```diff
@@ -49,19 +49,18 @@
 
     @Override
     public FetchSubPhaseProcessor getProcessor(SearchContext context) throws IOException {
+        List<FieldAndFormat> docValueFields = new ArrayList<>();
         if (context.collapse() != null) {
             // retrieve the `doc_value` associated with the collapse field
-            String name = context.collapse().getFieldName();
-            if (context.docValuesContext() == null) {
-                context.docValuesContext(new FetchDocValuesContext(
-                    Collections.singletonList(new FieldAndFormat(name, null))));
-            } else if (context.docValuesContext().fields().stream().map(ff -> ff.field).anyMatch(name::equals) == false) {
-                context.docValuesContext().fields().add(new FieldAndFormat(name, null));
-            }
+            docValueFields.add(new FieldAndFormat(context.collapse().getFieldName(), null));
```

Comment:
Can you keep the check that the `_collapse` field is not already requested ? Alternatively you could change the `docValueFields` List into a Set to make the deduplication automatic.